<a href="https://colab.research.google.com/github/oktsadwikar/DPDM2023/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
path = '/content/drive/MyDrive/DPDM23_DATA'

In [4]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [5]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [6]:
data.groupby('InvoiceNo').max()
# group the row by column Invoice number, and select the max value from each different column to be shown in the table

StockCode                       Description  Quantity  \
InvoiceNo                                                         
536365       85123A               WHITE METAL LANTERN         8   
536366        22633            HAND WARMER UNION JACK         6   
536367        84969       RECIPE BOX WITH METAL HEART        32   
536368        22960    YELLOW COAT RACK PARIS FASHION         6   
536369        21756          BATH BUILDING BLOCK WORD         3   
...             ...                               ...       ...   
C581484       23843       PAPER CRAFT , LITTLE BIRDIE    -80995   
C581490       23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
C581499           M                            Manual        -1   
C581568       21258        VICTORIAN SEWING BOX LARGE        -5   
C581569       84978  HANGING HEART JAR T-LIGHT HOLDER        -1   

                InvoiceDate  UnitPrice  CustomerID         Country  
InvoiceNo                                                           
536365     12/01/2010 08:26       7.65     17850.0  United Kingdom  
536366     12/01/2010 08:28       1.85     17850.0  United Kingdom  
536367     12/01/2010 08:34       9.95     13047.0  United Kingdom  
536368     12/01/2010 08:34       4.95     13047.0  United Kingdom  
536369     12/01/2010 08:35       5.95     13047.0  United Kingdom  
...                     ...        ...         ...             ...  
C581484    12/09/2011 09:27       2.08     16446.0  United Kingdom  
C581490    12/09/2011 09:57       1.95     14397.0  United Kingdom  
C581499    12/09/2011 10:28     224.69     15498.0  United Kingdom  
C581568    12/09/2011 11:57      10.95     15311.0  United Kingdom  
C581569    12/09/2011 11:58       1.25     17315.0  United Kingdom  

[25900 rows x 7 columns]

In [7]:
data.groupby('StockCode').max()
# group the row by column StockCode, and select the max value from each different column to be shown in the table

<ipython-input-7-abe775a9b6e7>:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.max is deprecated. In a future version, a TypeError will be raised. Before calling .max, select only columns which should be valid for the function.
  data.groupby('StockCode').max()


InvoiceNo  Quantity       InvoiceDate  UnitPrice  CustomerID  \
StockCode                                                                   
10002           551429       180   4/28/2011 15:05       1.66     18079.0   
10080           577801       170    9/27/2011 9:24       0.85     18096.0   
10120           580502        30   9/28/2011 15:18       0.21     18043.0   
10123C          560213         3   7/15/2011 15:05       0.65     17967.0   
10123G          549349       -38  04/08/2011 11:13       0.00         NaN   
...                ...       ...               ...        ...         ...   
gift_0001_20    573585        10   6/24/2011 15:51      17.02         NaN   
gift_0001_30    565231        30   8/30/2011 10:48      25.53         NaN   
gift_0001_40    561513         1   7/27/2011 15:12      34.04         NaN   
gift_0001_50    558614         1   6/30/2011 15:56      42.55         NaN   
m               539736         1  12/21/2010 15:18       2.55         NaN   

                     Country  
StockCode                     
10002         United Kingdom  
10080         United Kingdom  
10120         United Kingdom  
10123C        United Kingdom  
10123G        United Kingdom  
...                      ...  
gift_0001_20  United Kingdom  
gift_0001_30  United Kingdom  
gift_0001_40  United Kingdom  
gift_0001_50  United Kingdom  
m             United Kingdom  

[4070 rows x 6 columns]

In [8]:
data.groupby('Country').count()
# group the row by column country, and want to show the count or number of value value from each column to be shown

InvoiceNo  StockCode  Description  Quantity  \
Country                                                             
Australia                  1259       1259         1259      1259   
Austria                     401        401          401       401   
Bahrain                      19         19           19        19   
Belgium                    2069       2069         2069      2069   
Brazil                       32         32           32        32   
Canada                      151        151          151       151   
Channel Islands             758        758          758       758   
Cyprus                      622        622          622       622   
Czech Republic               30         30           30        30   
Denmark                     389        389          389       389   
EIRE                       8196       8196         8196      8196   
European Community           61         61           61        61   
Finland                     695        695          695       695   
France                     8557       8557         8557      8557   
Germany                    9495       9495         9495      9495   
Greece                      146        146          146       146   
Hong Kong                   288        288          288       288   
Iceland                     182        182          182       182   
Israel                      297        297          297       297   
Italy                       803        803          803       803   
Japan                       358        358          358       358   
Lebanon                      45         45           45        45   
Lithuania                    35         35           35        35   
Malta                       127        127          127       127   
Netherlands                2371       2371         2371      2371   
Norway                     1086       1086         1086      1086   
Poland                      341        341          341       341   
Portugal                   1519       1519         1519      1519   
RSA                          58         58           58        58   
Saudi Arabia                 10         10           10        10   
Singapore                   229        229          229       229   
Spain                      2533       2533         2533      2533   
Sweden                      462        462          462       462   
Switzerland                2002       2002         2002      2002   
USA                         291        291          291       291   
United Arab Emirates         68         68           68        68   
United Kingdom           495478     495478       494024    495478   
Unspecified                 446        446          446       446   

                      InvoiceDate  UnitPrice  CustomerID  
Country                                                   
Australia                    1259       1259        1259  
Austria                       401        401         401  
Bahrain                        19         19          17  
Belgium                      2069       2069        2069  
Brazil                         32         32          32  
Canada                        151        151         151  
Channel Islands               758        758         758  
Cyprus                        622        622         622  
Czech Republic                 30         30          30  
Denmark                       389        389         389  
EIRE                         8196       8196        7485  
European Community             61         61          61  
Finland                       695        695         695  
France                       8557       8557        8491  
Germany                      9495       9495        9495  
Greece                        146        146         146  
Hong Kong                     288        288           0  
Iceland                       182        182         182  
Israel                        297        297         250  
Italy                 

In [9]:
data.groupby('InvoiceNo')['StockCode'].apply(list)
# Group the row by column InvoiceNo, and combine the stock code & show it

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [10]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [11]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5955 sha256=67a074d69867f9afc55fc2308731f2444043f417d85f56ec407c4ed92335ee33
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [12]:
from apyori import apriori

In [18]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [19]:
len(frequentItemsets)

4

In [20]:
frequentItemsets[0][0]

frozenset({'22423'})

In [21]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [22]:
frequentItemsets[0][0].__len__()

1

In [25]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'85099B', '22386'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])

In [35]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER', 'ROSES REGENCY TEACUP AND SAUCER '}

In [30]:
set(data[data['StockCode']=='22697']['Description'])

{'GREEN REGENCY TEACUP AND SAUCER'}

RelationRecord(items=frozenset({'22699', '22697'}),

support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [36]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'569415',
 'C558255',
 '570258',
 '558119',
 '574076',
 '566255',
 '548860',
 '568655',
 '579267',
 '559510',
 '554959',
 '544887',
 '538177',
 '573320',
 '545696',
 '563896',
 '552614',
 '542087',
 '557431',
 '553718',
 '565724',
 '556237',
 '561702',
 '540662',
 '565748',
 '550210',
 'C560588',
 '537666',
 '537042',
 '565862',
 '550638',
 '546498',
 '566976',
 '560976',
 '554054',
 '562608',
 '564835',
 '537313',
 '540981',
 '558073',
 '552149',
 '572028',
 '579512',
 '546007',
 '573125',
 '568676',
 '547841',
 '570374',
 '573892',
 '581345',
 '548194',
 '543107',
 '561894',
 '581498',
 '558030',
 '549730',
 '546892',
 '542344',
 '567673',
 '578354',
 '543530',
 '548714',
 '576388',
 '549587',
 '546683',
 '536980',
 '559886',
 '546643',
 'C566758',
 '549295',
 '544305',
 '553612',
 '555278',
 '554329',
 '549222',
 '546008',
 '575607',
 '558262',
 '538371',
 '550125',
 '555743',
 '557859',
 '565286',
 '577076',
 '562417',
 '557605',
 '553542',
 '556826',
 '541423',
 '550519',
 '54037

In [37]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom